text/0032-improved-api-tokens.md

```diff
@@ -0,0 +1,153 @@
+
+### Token Format
+
+We need a predictable token format in order to integrate properly with secret scanning services. Our current format is a 64 character alphanumeric string. This is insufficient and would likely produce a high amount of false positives in tooling like [TruffleHog](https://github.com/trufflesecurity/trufflehog), [detect-secrets](https://github.com/Yelp/detect-secrets), [Github's Secret Scanning](https://docs.github.com/en/code-security/secret-scanning/about-secret-scanning), etc.
+
+The suggested pattern is `snty[a-zA-Z]_[a-zA-Z0-9]{64}`. The character _after_ `snty` will be used to identify the token type.
+
+For example:
+
+- `sntyu_a1b2c3...` - identifies a **user** API token
+- `sntya_a1b2c3...` - identifies an **API Application** token
+
+### Backend Token Hygiene
+
+API tokens should be stored in the database with a cryptographically secure hash (minimum SHA-256). Salting is not necessary since these values are sufficient in length.
+
+An additional column identifying the API token version will be added in order to correctly identify legacy API tokens.
+
+### Secret Verification Service
+
+In order to integrate with Github's Secret Scanning service, we will need to have an available service for them to submit potential matches to.
+
+More details about Github's secret scanning program can be found [here](https://docs.github.com/en/developers/overview/secret-scanning-partner-program).
+
+At a high-level, we will need to be able to:
+
+- Receive a list of one or more potential tokens
+- [Verify the signature of the request](https://docs.github.com/en/developers/overview/secret-scanning-partner-program#implement-signature-verification-in-your-secret-alert-service)
+- Determine if the token is a true positive or a false positive
+  - send this feedback back to Github
+- Disable/revoke the compromised token
+- Send a notification to the appropriate user
+
+> :memo: This secret verification service is not an immediate need or a dependency of implementing the improved API tokens.
+
+# Background
+
+Accidental credential leaks [happen](https://github.com/ChALkeR/notes/blob/master/Do-not-underestimate-credentials-leaks.md). Even though we provide several tools for users to limit the storage of sensitive information in Sentry it can still happen.
+
+Our support for various authentication methods (including 2FA, SSO, and SAML) helps mitigate access of potentially sensitive information, but API tokens cannot support these additional authentication gates.
+
+We can help customers further protect their accounts and data by providing a means of auto-detecting leaked API tokens.
+
+# Options Considered
+
+## Option #1
+
+This option strives to limit impact to the database by avoiding large bulk operations. To do this we'll distribute the complete implementation over two features. This also allows for a smooth transition of self-hosted instances keeping pace with releases.
+
+**In the first release:**
```

Comment:
I would have expected another, prior, backend-only release that ensures the APIs are forward-backward compatible with these changes.

Reply:
Apologies, I had a typo here. It would be distributed "over two releases". For the sake of this PR (and maybe I should clarify this in the document somewhere), I just looking at the major releases, for example, 22.11.0 and 22.12.0. This doesn't mean a single PR for each release, but enough incremental changes and PRs to make sure we do this in a backwards compatible way.